test/test_alloc.c

```diff
@@ -0,0 +1,401 @@
+/*
+ * Copyright (c) The mlkem-native project authors
+ * SPDX-License-Identifier: Apache-2.0 OR ISC OR MIT
+ */
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+
+/* Expose declaration of allocator (normally internal) */
+#define MLK_BUILD_INTERNAL
+#include "../mlkem/mlkem_native.h"
+#include "notrandombytes/notrandombytes.h"
+
+/*
+ * This test checks that
+ * - we handle allocator failures correctly, propagating MLK_ERR_OUT_OF_MEMORY
+ *   and cleaning up all memory, and
+ * - we leak no memory, and
+ * - we always de-allocate in the reverse order of allocation, thereby
+ *   allowing the use of a bump allocator.
+ *
+ * This is done through a custom bump allocator and tracking of in-flight
+ * allocations.
+ */
+
+/*
+ * Allocation tracker
+ *
+ * Simple stack of in-flight allocations that's used to test that there are
+ * no leaks and that we free in reverse order than we allocate. (The absence
+ * of leaks is also checked through the address sanitizer)
+ */
+
+typedef struct
+{
+  void *addr;
+  size_t size;
+  const char *file;
+  int line;
+  const char *var;
+  const char *type;
+} alloc_info_t;
+
+#define MLK_MAX_IN_FLIGHT_ALLOCS 100
+static alloc_info_t alloc_stack[MLK_MAX_IN_FLIGHT_ALLOCS];
+static int alloc_stack_top = 0;
+
+static void alloc_tracker_push(void *addr, size_t size, const char *file,
+                               int line, const char *var, const char *type)
+{
+  if (alloc_stack_top >= MLK_MAX_IN_FLIGHT_ALLOCS)
+  {
+    fprintf(stderr, "ERROR: Allocation stack overflow\n");
+    exit(1);
+  }
+  alloc_stack[alloc_stack_top].addr = addr;
+  alloc_stack[alloc_stack_top].size = size;
+  alloc_stack[alloc_stack_top].file = file;
+  alloc_stack[alloc_stack_top].line = line;
+  alloc_stack[alloc_stack_top].var = var;
+  alloc_stack[alloc_stack_top].type = type;
+  alloc_stack_top++;
+}
+
+static void alloc_tracker_pop(void *addr, const char *file, int line,
+                              const char *var)
+{
+  alloc_info_t *top;
+  if (alloc_stack_top == 0)
+  {
+    fprintf(
+        stderr,
+        "ERROR: Attempting to free %s at %s:%d but allocation stack is empty\n",
+        var, file, line);
+    exit(1);
+  }
+
+  top = &alloc_stack[alloc_stack_top - 1];
+  if (top->addr != addr)
```

Comment:
Can you extend this to also error if the wrong size is passed to MLK_FREE? That is currently not detected.